src/full/Agda/TypeChecking/Monad/Base.hs

```diff
@@ -2677,12 +2728,491 @@ ifTopLevelAndHighlightingLevelIs ::
+
+infixr 5 ⊣::
+(⊣::) :: HasPids a => [a] -> ContextHet' a -> ContextHet' a
+[]     ⊣:: γΓ = γΓ
+(b:bs) ⊣:: γΓ = Entry (getPids b <> getPids γΓ') (b :∈ γΓ')
+  where γΓ' = bs ⊣:: γΓ
+
+(⊣!::) :: (AsTwin a, AsTwin_ a ~ b) => [b] -> ContextHet' a -> ContextHet' a
+[]     ⊣!:: γΓ = γΓ
+(b:bs) ⊣!:: γΓ = Entry (getPids γΓ) (asTwin b :∈ (bs ⊣!:: γΓ))
+
+
+-- * Manipulating context as a list.
+
+-- | > contextHetAsList (a :⊣ b :⊣ … :⊣ Empty)
+--     (a:b:…:[])
+-- Also: contextHetToList :: ContextHet -> [Dom (Name, TwinT)]
+contextHetToList :: Context_' a -> [a]
+contextHetToList Empty = []
+contextHetToList (a :⊣ γΓ) = a:contextHetToList γΓ
+
+contextHetFromList :: HasPids a => [a] -> Context_' a
+contextHetFromList []     = Empty
+contextHetFromList (a:as) = a ⊣: contextHetFromList as
+
+-- * Switch heterogeneous context to a specific side
+switchSide :: forall s a m. (HetSideIsType s, MonadTCEnv m) => m a -> m a
+switchSide = unsafeModifyContext {- IdS -} (asTwin . twinAt @s)
+
+{-# INLINE switchSide_ #-}
+switchSide_ :: forall s a m. (HetSideIsType s, MonadTCEnv m) =>
+  ((forall b n. MonadTCEnv n => n b -> n b) -> m a) -> m a
+switchSide_ κ = do
+  ctx <- asksTC envContext
+  unsafeModifyContext {- IdS -} (asTwin . twinAt @s) (κ (unsafeModifyContext {- IdS -} (const ctx)))
+
+-- * Heterogeneous telescope
+
+type Telescope_ = Tele (Dom TwinT)
+
+-- * Describing parts of twin contexts
+
+data HetSide = LHS | RHS | Compat | Both
+
+type H'LHS = Het 'LHS
+type H'RHS = Het 'RHS
+type H'Compat = Het 'Compat
+type H'Both = Het 'Both
+
+pattern H'LHS :: a -> H'LHS a
+pattern H'LHS a = Het a
+
+pattern H'RHS :: a -> H'RHS a
+pattern H'RHS a = Het a
+
+pattern H'Compat :: a -> H'Compat a
+pattern H'Compat a = Het a
+
+pattern H'Both :: a -> H'Both a
+pattern H'Both a = Het a
+
+#if __GLASGOW_HASKELL__ >= 804
+{-# COMPLETE H'LHS #-}
+{-# COMPLETE H'RHS #-}
+{-# COMPLETE H'Compat #-}
+{-# COMPLETE H'Both #-}
+#endif
+
+type OnSide = Het
+type OnLHS = OnSide 'LHS
+type OnRHS = OnSide 'RHS
+
+pattern OnLHS :: a -> OnLHS a
+pattern OnLHS a = Het a
+pattern OnRHS :: a -> OnRHS a
+pattern OnRHS a = Het a
+
+#if __GLASGOW_HASKELL__ >= 804
+{-# COMPLETE OnLHS #-}
+{-# COMPLETE OnRHS #-}
+#endif
+
+onSide :: forall s a m b. (MonadAddContext m, HetSideIsType s) => (a -> m b) -> Het s a -> m (Het s b)
+onSide κ = switchSide @s . traverse κ
+
+onSide_ :: forall s a m b. (MonadAddContext m, HetSideIsType s) => (a -> m b) -> Het s a -> m b
+onSide_ κ = fmap (unHet @s) . switchSide @s . traverse κ
+
+-- Dependent type boilerplate
+data instance SingT (a :: HetSide) where
+  SLHS    :: SingT 'LHS
+  SRHS    :: SingT 'RHS
+  SCompat :: SingT 'Compat
+  SBoth   :: SingT 'Both
+instance Sing 'LHS    where sing = SLHS
+instance Sing 'RHS    where sing = SRHS
+instance Sing 'Both   where sing = SBoth
+instance Sing 'Compat where sing = SCompat
+
+-- | Distinguishes which sides of a twin type corresponds to a single type
+type family HetSideIsType_ (s :: HetSide) :: Bool where
```

Comment:
What's the purpose of this `const True` type family?

Reply:
Fixed; replaced it by `SideIsSingle`.
The purpose is to distinguish terms that live on one side from terms that live in both of them.